runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -2140,6 +2211,124 @@ J9::ValuePropagation::transformFlattenedArrayElementStore(TR_OpaqueClassBlock *a
    return false;
    }
 
+void
+J9::ValuePropagation::transformFlattenedArrayElementLoadStoreUseTypeHint(TR_OpaqueClassBlock *typeHintClass, TR::Node *callNode, TR::TreeTop *callTree, bool isLoad, bool needsNullValueCheck)
+   {
+   /*
+    * Use the transformation of jitLoadFlattenableArrayElement as an example below
+    *
+    * public primitive class SomeClass implements SomeIf {
+    *     public final double d;
+    *     public final int i;
+    * }
+    *
+    * Trees before Global Value Propagation:
+    * n27n      treetop
+    * n26n        acall  jitLoadFlattenableArrayElement
+    * n23n          iload  <auto slot 2>[#420  Auto]
+    * n22n          aload  <auto slot 1>[#419  Auto]
+    * n89n      treetop after the original jitLoadFlattenableArrayElement
+    *
+    * Trees after Global Value Propagation:
+    * n96n      istore  <temp slot 4>[#428  Auto]
+    * n23n        iload  <auto slot 2>[#420  Auto]
+    * n98n      astore  <temp slot 5>[#429  Auto]
+    * n22n        aload  <auto slot 1>[#419  Auto]
+    * n102n     ificmpne --> block_14 BBStart at n114n ()
+    * n101n       aloadi  <vft-symbol>[#345  Shadow]
+    * n99n          aload  <temp slot 5>[#429  Auto]
+    * 100n       aconst 0x1ccb00 Abstract ([L.class) (classPointerConstant sharedMemory )
+    * n113n     BBEnd </block_4> =====
+
+    * n117n     BBStart <block_15>
+    * n107n     treetop
+    * n108n       newvalue  jitNewValue
+    * n121n         loadaddr  SomeClass
+    * n128n         iloadi  <Q-SomeClass-array-shadow>.i I[#432  <Q-SomeClass-array-shadow>.i I]
+    * n127n           aladd (internalPtr sharedMemory )
+                ...
+    * n129n         dloadi  <Q-SomeClass-array-shadow>.d D[#433  <Q-SomeClass-array-shadow>.d D +4]
+    * n127n           ==>aladd
+    * n120n     astore  <temp slot 6>[#430  Auto]
+    * n108n       ==>newvalue
+    * n118n     BBEnd </block_15> =====
+
+    * n112n     BBStart <block_13>
+    * n80n      astore  <temp slot 2>[#426  Auto]
+    * n26n        aload  <temp slot 6>[#430  Auto]
+    * n89n      treetop after the original jitLoadFlattenableArrayElement
+    * ...
+    * n82n      BBEnd </block_13> =====
+    * ...
+    * n114n     BBStart <block_14>
```

Comment:
I think block_14 will be inserted at the end of the method, right?  You might want to clarify that here.

Reply:
Addressed in 3b5dc03